include/wil/rpc_helpers.h

```diff
@@ -0,0 +1,170 @@
+    within the RPC machinery. This method wraps the requested call in the usual RpcTryExcept,
+    RpcTryCatch, and RpcEndExcept sequence then maps the exceptions to HRESULTs for the usual
+    flow control machinery to use.
+
+    Many RPC methods are defined as returning HRESULT themselves, where the HRESULT indicates
+    the result of the _work_. HRESULTs returned by a successful completion of the _call_ are
+    returned as-is.
+
+    RPC methods that have a return type of 'void' are mapped to returning S_OK when the _call_
+    completes successfully.
+
+    For example, consider an RPC interface method defined in idl as:
+    ~~~
+    HRESULT GetKittenState([in, ref, string] const wchar_t* name, [out, retval] KittenState** state);
+    ~~~
+    To call this method, use:
+    ~~~
+    wil::unique_rpc_binding binding = // typically gotten elsewhere;
+    wil::unique_midl_ptr<KittenState> state;
+    HRESULT hr = wil::invoke_rpc_nothrow(GetKittenState, binding.get(), L"fluffy", state.put());
+    RETURN_IF_FAILED(hr);
+    ~~~
+    */
+    template<typename... TCall> HRESULT invoke_rpc_nothrow(TCall&&... args) WI_NOEXCEPT
+    {
+        RpcTryExcept
+        {
+            // Note: this helper type can be removed with C++17 enabled via
+            // 'if constexpr(wistd::is_same_v<void, result_t>)'
+            using result_t = typename wistd::__invoke_of<TCall...>::type;
+            RETURN_IF_FAILED(details::call_adapter<result_t>::call(wistd::forward<TCall>(args)...));
+            return S_OK;
+        }
+        RpcExcept(RpcExceptionFilter(RpcExceptionCode()))
+        {
+            RETURN_HR(details::map_rpc_exception(RpcExceptionCode()));
+        }
+        RpcEndExcept
+    }
+
+    /** Invokes an RPC method, mapping structured exceptions to HRESULTs
+    Failures encountered by the RPC infrastructure (such as server crashes, authentication
+    errors, client parameter issues, etc.) are emitted by raising a structured exception from
+    within the RPC machinery. This method wraps the requested call in the usual RpcTryExcept,
+    RpcTryCatch, and RpcEndExcept sequence then maps the exceptions to HRESULTs for the usual
+    flow control machinery to use.
+
+    Some RPC methods return results (such as a state enumeration or other value) directly in
+    their signature. This adapter writes that result into a caller-provided object then
+    returns S_OK.
+
+    For example, consider an RPC interface method defined in idl as:
+    ~~~
+    GUID GetKittenId([in, ref, string] const wchar_t* name);
+    ~~~
+    To call this method, use:
+    ~~~
+    wil::unique_rpc_binding binding = // typically gotten elsewhere;
+    GUID id;
+    HRESULT hr = wil::invoke_rpc_result_nothrow(id, GetKittenId, binding.get(), L"fluffy");
+    RETURN_IF_FAILED(hr);
+    ~~~
+    */
+    template<typename TResult, typename... TCall> HRESULT invoke_rpc_result_nothrow(TResult& result, TCall&&... args) WI_NOEXCEPT
```

Comment:
If the `TCall...` requires conversions to match the arguments of the invoked function, those conversions will take place **inside** the `RpcTryExcept`, and if the conversions encounter C++ exceptions, crazy things will happen: The `RpcExceptionFilter` will say "Sure, I'll handle that exception" and the function will simply return `0xE06d7363` instead of propagating the C++ exception.

We need to perform the conversions outside the `RpcTryExcept` so the RPC exception handler won't swallow them.

One option is to convert it ourselves. This means that if an exception does occur, we will `std::terminate` because this function is marked `noexcept`.

```cpp
template<typename TResult, typename... TExpectedArgs, typename... TActualArgs>
HRESULT invoke_rpc_result_nothrow(
    TResult& result,
    TResult(__stdcall*function)(TExpectedArgs...),
    TActualArgs... args)
{
    // C++ exception possible during conversion of arguments
    std::tuple<TExpectedArgs...> converted{ std::forward<TActualArgs>(args)... };
    RpcTryExcept
    {
        result = std::apply(function, std::move(converted));
        return S_OK;
    }
    RpcExcept(RpcExceptionFilter(RpcExceptionCode()))
    {
        RETURN_HR(details::map_rpc_exception(RpcExceptionCode()));
    }
    RpcEndExcept
}
```

Another option is to disallow conversions in the arguments and require the caller to preconvert them. This allows us to remain `WI_NOEXCEPT`, and might even be more in the spirit of WIL: If you want the `nothrow` version, you can't do things that might throw.

```cpp
template<typename TResult, typename... TArgs>
HRESULT invoke_rpc_result_nothrow(
    TResult& result,
    TResult(__stdcall*function)(TArgs...),
    TArgs... args) WI_NOEXCEPT
{
    RpcTryExcept
    {
        result = wistd::invoke(wistd::forward<TCall>(args)...);
        return S_OK;
    }
    RpcExcept(RpcExceptionFilter(RpcExceptionCode()))
    {
        RETURN_HR(details::map_rpc_exception(RpcExceptionCode()));
    }
    RpcEndExcept
}
```

It does create extra work on the caller to do the preconversion.

```cpp
extern HRESULT GetMuffin(IMuffin** muffin);

wil::com_ptr_nothrow<IMuffin> muffin;
// does not compile
invoke_rpc_result_nothrow(hr, GetMuffin, wil::out_param(muffin));

// you have to preconvert, which is cumbersome
invoke_rpc_result_nothrow(hr, GetMuffin, static_cast<IMuffin**>(wil::out_param(muffin)));
```

Reply:
Filed #59 to resolve.